cpp/src/randomforest/randomforest.cu

```diff
@@ -241,79 +359,99 @@ rfClassifier<T>::rfClassifier(RF_params cfg_rf_params)
-                   (unsigned int)n_rows, stream);
-    } else {
-      std::vector<unsigned int> h_selected_rows(n_rows);
-      std::iota(h_selected_rows.begin(), h_selected_rows.end(), 0);
-      std::random_shuffle(h_selected_rows.begin(), h_selected_rows.end());
-      h_selected_rows.resize(n_sampled_rows);
-      MLCommon::updateDevice(selected_rows.data(), h_selected_rows.data(),
-                             n_sampled_rows, stream);
-    }
+    this->prepare_fit_per_tree(handle, i, n_rows, n_sampled_rows,
+                               selected_rows.data(),
+                               sorted_selected_rows.data(),
+                               rows_temp_storage->data(), temp_storage_bytes);
 
     /* Build individual tree in the forest.
 		   - input is a pointer to orig data that have n_cols features and n_rows rows.
 		   - n_sampled_rows: # rows sampled for tree's bootstrap sample.
-		   - selected_rows: points to a list of row #s (w/ n_sampled_rows elements) used to build the bootstrapped sample.
-			Expectation: Each tree node will contain (a) # n_sampled_rows and (b) a pointer to a list of row numbers w.r.t original data.
+		   - sorted_selected_rows: points to a list of row #s (w/ n_sampled_rows elements) used to build the bootstrapped sample.
+		   Expectation: Each tree node will contain (a) # n_sampled_rows and (b) a pointer to a list of row numbers w.r.t original data.
 		*/
-    this->trees[i].fit(user_handle, input, n_cols, n_rows, labels,
-                       selected_rows.data(), n_sampled_rows, n_unique_labels,
-                       this->rf_params.tree_params);
 
-    //Cleanup
-    selected_rows.release(stream);
+    trees[i].fit(user_handle, input, n_cols, n_rows, labels,
+                 sorted_selected_rows.data(), n_sampled_rows, n_unique_labels,
+                 this->rf_params.tree_params, tempmem);
   }
+
+  //Cleanup
+  rows_temp_storage->release(stream);
+  selected_rows.release(stream);
+  sorted_selected_rows.release(stream);
+  tempmem.reset();
+  delete rows_temp_storage;
 }
 
 /**
  * @brief Predict target feature for input data; n-ary classification for single feature supported.
  * @tparam T: data type for input data (float or double).
- * @param[in] user_handle: cumlHandle (currently unused; API placeholder)
- * @param[in] input: test data (n_rows samples, n_cols features) in row major format. CPU pointer.
+ * @param[in] user_handle: cumlHandle.
+ * @param[in] input: test data (n_rows samples, n_cols features) in row major format. GPU pointer.
  * @param[in] n_rows: number of  data samples.
  * @param[in] n_cols: number of features (excluding target feature).
- * @param[in, out] predictions: n_rows predicted labels. CPU pointer, user allocated.
+ * @param[in, out] predictions: n_rows predicted labels. GPU pointer, user allocated.
  * @param[in] verbose: flag for debugging purposes.
  */
 template <typename T>
 void rfClassifier<T>::predict(const cumlHandle& user_handle, const T* input,
```

Comment:
Can we add a version of predict function that returns all the estimators (DTs) predictions for every sample? Will use this function in dask RF for MNMG prediction. You can add this in a separate PR as well.

Reply:
You don’t really need predictions per tree. Rf gives max vote for classification and mean value for regression. The same thing can be re done for multiple dash rf. The hierarchical prediction would be the same as doing it all at same time..

Reply:
If we're talking about having subsets of trees, then the majority vote across all trees won't necessarily be the majority of the majority vote of each subset. 

If a new prediction function is needed for MNMG, we can certainly file a separate PR for this (e.g., even add it to the stateless API one).

Reply:
Yeah, I think this PR is huge; Lets add this is a new PR.